### ISSUE TITLE
4.8-Classification

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learing-NTU#34]
如果要做資料 Regression（預測）的話，要給予基本的 Model + Data，並且利用 Loss function（損失函數）來檢查現在的 function 的糟糕程度，並利用 Gradient Descent 更新 Model 來模擬窮舉的概念達到找到最佳函數的效果

分類問題會在什麼時候出現?判斷 yes/no 問題, 分辨不同的動物, 分辨人臉

________________________________
### 以天氣分類問題為例子

假設一天的天氣有分成三種，分別是晴天、陰天、雨天，所以相較於 Regression 輸出是一個實數（預測結果），Classification 的 Output 就會分別是晴天、陰天、雨天三種結果

回顧一下機器學習的流程，給 Model（function set） -> Goodness of function -> Find the best function

那我們說過 Goodness of function 可以利用 Loss function 來檢查當前 function 的錯誤程度，然後在最後輸出 best function
在 Regression loss function 可以直接用 Estimation error（估計誤差），然後更新參數用 Gradient Descent，那分類問題呢?

### Loss Function

在 Regression 中，我們的 Loss function 會是預期的答案減去我們計算出來的答案的差異量，但是 Classification 這邊並沒有答案差異這件事情，要怎麼辦呢?

我們平常怎麼知道自己的判斷力好不好？也就是看我們判斷的成功率

所以我們可以利用不符合的答案數量來當作回答狀況的好壞，所以我們可以改寫我們的 Loss function 成
![image](https://user-images.githubusercontent.com/34474924/212008496-2c96be61-80fc-4d00-a6b6-5354232ee450.png)

這邊很明顯不能用 Gradient Descent 去最佳化我們 Model 的參數，那有沒有別的方法可以解決這邊最佳化的問題嗎?
其實是有的，例如 Perceptron Learning Algorithm（PLA）， Support Vector Machine（SVM） 等等，那直接說明 PLA 或是 SVM 會很不直觀，因此我們換個方式解釋

我們先假設一個情境，今天我們的世界只有晴天 & 雨天兩種天氣，因此我們的工作就是判斷明天是晴天 OR 明天是雨天兩種問題
那我們再來假設一下，我們判斷兩種問題的資料也很簡單，就是溫度&濕度兩個變數(feature)，所以如果要把資料畫成圖, 應該怎麼判斷這些資料誰屬於誰？很簡單，那就是用機率來看

![image](https://user-images.githubusercontent.com/34474924/215441603-33dd0c35-3893-4217-b66d-a94fe210f0c0.png)

如果我們想要分開資料，在肉眼辨識下我們大致上會怎麼去切分?

![image](https://user-images.githubusercontent.com/34474924/215441707-d430ad61-10b9-44b7-b559-f22dec12c217.png)

_________________________

我們現在有兩個盒子，盒子一跟盒子二，這兩個盒子都有紅色球球跟綠色球球，今天如果我們拿到了一顆紅色的球球，我們只知道這個紅色球球是從這兩個盒子裡面拿到的，請問他是從盒子一拿出來的機率是多少?從盒子二拿出來的機率是多少?

我要算從第一個盒子取出紅色球的機率，就會是取第一個盒子的機率乘上第一個盒子裡面紅色球被拿到的機率去除與兩個盒子拿到紅色球的機率

![image](https://user-images.githubusercontent.com/34474924/212016659-ce3ba4a5-222f-4691-935b-cb80844286eb.png)

我們假設今天我們把盒子改成類別，然後給定一個 x，這個 x 是我們要分類的對象，也就是要分成 class1 或是 class2，那我們這時候先來算這個 x 出現在 class1 跟 class2 中的機率是多少，也就是我從這兩個 class 中抽取到 x 的機率

![image](https://user-images.githubusercontent.com/34474924/215441923-690bb546-592e-42d5-beae-9a7d448d83cf.png)

![image](https://user-images.githubusercontent.com/34474924/212016784-8ace77e5-011c-4d18-9f1a-8210782acdc3.png)

我們可以很簡單的知道這個天氣狀況到底比較偏向哪一種形況，他到底是晴天的機率比較大，還是雨天的機率比較大

那這邊我們等於需要去取得四個變數

![image](https://user-images.githubusercontent.com/34474924/212016908-59c2d4ef-4c80-4a9c-a043-f10f4a77e3ed.png)

我們只要有這四個變數，我們就能夠計算出任意一個 x 進來之後，他於兩個 cases 機率了，而現在這個做法叫做一個 Generative Model（生成模型），因為我只要能夠有這個模型，我就能生成所有資料點的 Distribution（分布），來判斷資料狀況。

高斯分布

![image](https://user-images.githubusercontent.com/34474924/212017428-8d32c1da-8f2c-49ec-868a-a519ed30341c.png)

這四個變數要怎麼生成，重點要先把P(C1)&P(C2)兩個機率確認，而這個機率怎麼確認，就是利用高斯分布來確認現在訓練的資料組裡面，的分布機率狀態

所以概念上就是只要能搞好高斯分布(找到好的∑和µ)，就能得到兩群出現的機率，就可以間接得到類別中後驗機率，然後就能夠計算它的機率了

_________________

### likelihood function（概率函數）

![image](https://user-images.githubusercontent.com/34474924/215442616-80eeee4b-4dba-4728-b53d-d8a361cd8742.png)

圖上的兩個紅圈圈分別代表著兩個不同的高斯分布中心，那我們可以很明顯知道這兩個高斯分布去計算對於晴天的分布狀況會很明顯地不一樣，比較偏中間的那個高斯分布的可能性會比比較偏左下的那個可能性機率低

其實就可以去算 likelihood 的狀況, ikelihood 越高理論上我們的高斯分布越好=好的∑和µ

likelihood function 
![image](https://user-images.githubusercontent.com/34474924/215442983-0e93bf41-6d7b-4dc0-9b41-a95003cc4e66.png)

[https://ithelp.ithome.com.tw/articles/10268944](https://ithelp.ithome.com.tw/articles/10268944)

________

###  Logistic Regression
披著 Regression 外皮的 Classification 

Logistic Regression 是一種分類模型，利用現有的資料點，盡可能的找出能將所有資料點分離的那條線

線性回歸是用來預測一個連續的值，羅吉斯回歸是用來分類

![image](https://user-images.githubusercontent.com/34474924/212018793-fb716112-4d72-40b5-8170-847d422e2800.png)

從圖上來看，線性回歸跟羅吉斯回歸公式是一樣的
線性回歸輸出是一個連續的實數，但羅吉斯回歸就是用線性回歸的輸出來判斷這個資料屬不屬於target(二分類問題: target和non-target)，但羅吉斯回歸要怎麼做分類?

一個最簡單的概念，將點帶進去回歸線，回歸線輸出值若是>=0，是一類(target)，值<0是另一類(non-target)

![image](https://user-images.githubusercontent.com/34474924/212028848-f37b3e83-5cc6-46ac-83e2-84bd2f44ef3b.png)

這個判斷法其實就是一個unit-step function，如下
![image](https://user-images.githubusercontent.com/34474924/212028897-ef38b761-f067-477a-8460-39d97514c6b6.png)

這種輸出很直覺也很直接，但演算法開發很不喜歡這樣的東西，因為太絕對了，越靠近0的位置，理論上越容易分類錯誤，所以這時候我們加上一個自然底數函數，這時候這個輸出就更有彈性。

羅吉斯回歸用到的對數函數是Sigmoid函數，如下(這邊寫的Sigmoid是兩個簡易版，因為網路這兩種寫法都有人寫，所以我都列出來，當然還有複雜版，谷哥一下會找到Sigmoid還有平移和放大縮小的參數):

![image](https://user-images.githubusercontent.com/34474924/212033787-9132da39-b638-412f-a468-38705d470f0f.png)

![image](https://user-images.githubusercontent.com/34474924/212033816-42ffb11a-56fc-48b6-afdf-048156e5aba1.png)

![image](https://user-images.githubusercontent.com/34474924/212033970-e529b152-fb54-4b83-a87b-41375740cb59.png)

因此 Logistic Regression 利用對數函數 Sigmoid 來達到更有彈性的輸出

一樣拉到三步驟去做比較1.Model 2.Goodness of function 3.Find the best function

_________________________
### Model

Linear Regression 的公式很簡單，就是
![image](https://user-images.githubusercontent.com/34474924/212260095-c3202a32-2397-4fb4-9091-9a0dfe89b8e0.png)

但是 Logistic Regression 前面有提到，是要分類的，因此公式上多了一個部分，就是σ
![image](https://user-images.githubusercontent.com/34474924/212260735-b4f8e51f-e42d-4349-a65b-406914b7c43d.png)
Output 會是 0 or 1
公式真的一模一樣，唯一差異就是 Logistic Regression 多了一個 Sigmoid，去幫忙分類 0,1

### Goodness of function
Linear Regression 就是一個很簡單的 Loss function，就是 Mean Square Error=也就是去看答案的差異
然而 Logistic Regression 無法直接用 Mean Square Error
 詳細的解釋-[李弘毅老師](https://youtu.be/hSXFuypLukA) 的開放式課程

假設我們去計算data1被分類成class1的機率是P w,b (C1|x)
那這個w,b好壞程度就是
![image](https://user-images.githubusercontent.com/34474924/212261639-b7068bad-ccc3-4b84-a480-f7f3a1df27b5.png)
之過經過一系列推倒就可以得到 Logistic Regression 的 Loss Function 了(Cross-entropy loss function)

### Find the best function
我們開始微分我們的 cross-entropy loss function，詳細的數學運算我們一樣跳過，結論就是我們一樣用 Gradient Descent 得到了
![image](https://user-images.githubusercontent.com/34474924/212262130-3ec9adfe-4453-4c50-886a-c15f43a51e6b.png)

這邊和當初 Linear Regression 的 Gradient Descent幾乎一樣，唯一的差別就是 Logistic Regression是介於 0~1 之間的數值

[https://chih-sheng-huang821.medium.com/%E6%A9%9F%E5%99%A8-%E7%B5%B1%E8%A8%88%E5%AD%B8%E7%BF%92-%E7%BE%85%E5%90%89%E6%96%AF%E5%9B%9E%E6%AD%B8-logistic-regression-aff7a830fb5d](https://chih-sheng-huang821.medium.com/%E6%A9%9F%E5%99%A8-%E7%B5%B1%E8%A8%88%E5%AD%B8%E7%BF%92-%E7%BE%85%E5%90%89%E6%96%AF%E5%9B%9E%E6%AD%B8-logistic-regression-aff7a830fb5d)

[https://ithelp.ithome.com.tw/articles/10269617](https://ithelp.ithome.com.tw/articles/10269617)
